File: aoc2024/day1b.py
import sys


def run():
    left = []
    right = []
    result = 0
    for line in sys.stdin:
        s = line.strip().split(" ")
        left.append(int(s[0]))
        right.append(int(s[-1]))

    counts = {}
    for r in right:
        counts[r] = counts.get(r, 0) + 1

    for l in left:
        result += counts.get(l, 0) * l

    print(result)


if __name__ == "__main__":
    run()

File: src/aoc2024/day1b.rs
use std::collections::HashMap;
use std::io::BufRead;

pub fn run() {
    let mut counts: HashMap<i32, i32> = HashMap::new();

    let lines: Vec<Vec<i32>> = std::io::stdin()
        .lock()
        .lines()
        .map(|x| {
            x.unwrap()
                .split_whitespace()
                .map(|y| y.parse::<i32>().unwrap())
                .collect()
        })
        .collect();

    let left: Vec<i32> = lines.iter().map(|x| *x.first().unwrap()).collect();
    lines.iter().for_each(|x| {
        let key = x.last().unwrap();
        match counts.get(key) {
            Some(count) => {
                counts.insert(*key, *count + 1);
            }
            None => {
                counts.insert(*key, 1);
            }
        }
    });

    let result: i32 = left
        .into_iter()
        .map(|x| match counts.get(&x) {
            Some(i) => x * i,
            None => 0,
        })
        .sum();

    println!("{result}")
}
